rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/sync/qlearning/QLearning.java

```diff
@@ -136,49 +139,47 @@ protected StatEntry trainEpoch() {
             reward += stepR.getStepReply().getReward();
             obs = stepR.getStepReply().getObservation();
             incrementStep();
-            step++;
         }
 
         meanQ /= (numQ + 0.001); //avoid div zero
 
 
-        StatEntry statEntry = new QLStatEntry(getStepCounter(), getEpochCounter(), reward, step, scores,
+        StatEntry statEntry = new QLStatEntry(getStepCounter(), getCurrentEpochStep(), reward, currentEpochStep, scores,
```

Comment:
Were those two always effectively equal?

Reply:
I mean, it looks like the number of epochs and the number of steps have been confounded here and in the rest of the changes. Could you make sure this isn't the case?

Reply:
You're right, this is wrong. There are other places where I mixed epoch step and epoch number. I'll fix this asap. Really sorry!

Reply:
Done.

Reply:
Do you think we still need StepCountable.java? Or can we replace it with the new interfaces?

Reply:
That `StepCountable`/`stepCounter` is bothering me too. The way steps are handled is a mess. I think that in the async case, it's not even the right count. Presently, it's the total number of steps per thread (not reset between episodes but not shared between threads). But in literature, the closest concept is the number of iterations which is the grand total of all steps of all episodes taken by all agents. 

We shouldn't even need an interface. I think this problem is part of a bigger one. The learning classes and `Policy.play()` are not simply doing something similar; they all have their separate implementation of what is an agent. And this agent concept is missing in RL4J.

I am still thinking on the details but there should be a separate Agent class. The Policy shouldn't be used to `play()` and the learning classes should contain an agent instead of being one. And specific behaviors should be injected instead of being implemented through inheritance.

Reply:
I have removed StepCountable. You are right, lets remove it now since it's not used.